api/authentication.py

```diff
@@ -10,77 +9,99 @@
             return None
 
         token = authorization.split(" ")[1]
         cache_key = get_cache_key(token)
-        cached_fxa_resp_data = fxa_resp_data = cache.get(cache_key)
         # set a default cache_timeout, but this will be overriden to match
-        # the 'exp' time returned by FXA
+        # the 'exp' time in the JWT returned by FXA
         cache_timeout = 60
+        cached_fxa_resp_data = fxa_resp_data = cache.get(cache_key)
         if not fxa_resp_data:
+            # set a default fxa_resp_data, so any error during introspection
+            # will still cache for at least cache_timeout to prevent an outage
+            # from causing useless run-away repetitive introspection requests
+            fxa_resp_data = {"status_code": None, "json": {}}
             introspect_token_url = (
                 "%s/introspect"
                 % settings.SOCIALACCOUNT_PROVIDERS["fxa"]["OAUTH_ENDPOINT"]
             )
-            fxa_resp = requests.post(introspect_token_url, json={"token": token})
-            fxa_resp_data = {"status_code": fxa_resp.status_code, "json": None}
             try:
-                fxa_resp_data["json"] = fxa_resp.json()
-            except requests.exceptions.JSONDecodeError:
-                logger.error(
-                    "JSONDecodeError from FXA introspect response.",
-                    extra={"fxa_response": shlex.quote(fxa_resp.text)},
-                )
+                fxa_resp_data = introspect_token(introspect_token_url, token)
+            except AuthenticationFailed:
+                raise
+            finally:
+                cache.set(cache_key, fxa_resp_data, cache_timeout)
 
         user = None
-        if (
-            fxa_resp_data
-            and fxa_resp_data["status_code"] == 200
-            and fxa_resp_data["json"]
-            and fxa_resp_data["json"].get("active")
-        ):
-
-            # FxA user is active, check for the associated Relay account
-            fxa_uid = fxa_resp_data.get("json").get("sub")
-            if fxa_uid:
-                try:
-                    sa = SocialAccount.objects.get(uid=fxa_uid, provider="fxa")
-                except SocialAccount.DoesNotExist:
-                    # No Relay account associated with the FxA ID. It might be
-                    # a user who deleted their Relay account since they first
-                    # signed up
-                    pass
-                else:
-                    user = sa.user
-
-                    # cache fxa_resp_data for as long as access_token is valid
-                    # Note: FXA iat and exp are timestamps in *milliseconds*
-                    fxa_token_exp_time = int(
-                        fxa_resp_data.get("json").get("exp") / 1000
-                    )
-                    now_time = int(datetime.now(timezone.utc).timestamp())
-                    cache_timeout = fxa_token_exp_time - now_time
+        if not fxa_resp_data["status_code"] == 200:
```

Comment:
What if we add this before the 200 check?

```python
if fxa_resp_data["status_code"] is None:
   raise AuthenticationFailed("Previous FXA call failed, wait to retry")
```

That way we'll get a slightly different response for a cached FXA failure.

Also, I always type it `FxA` not `FXA`. What will Mozilla Accounts be?